### PR TITLE
Improve deeplink tracking

### LINKED ISSUE #14647: Improve deeplink tracking
The current tracks contain the full clicked URL. We want to strip it and make sure tracked data makes sense.

### PR DESCRIPTION
**Pull request by planarvoid**
Fixes #14647 

I'm sorry for the size of this PR. It unexpectedly grew over the time with the changing requirements. What it contains is:
- strip the URL tracked with each deep link click of any data containing any identifiers
- add `source` to each event - tracking where the link came from - email, banner or link
- add `source_info` to each event - additional information - in our case only `login_reason` from the `create_site` links

In order to achieve this I've extracted a `DeepLinkHandler` interface and create a provider for all the handlers. I think the resulting structure is neat and simplifies the `ViewModel` a lot. Let me know what you think! I've tried to cover everything with unit tests.

For the `read` URLs from the `ReaderLinkHandler` that are now redirected to the `ReaderPostPagerActivity` - these are now tracked twice. I think that's the cleanest approach at the moment. Both of these have a different `action` so it should be easy to differentiate between these two events. Let me know what you think about this approach!

To test:
- Try various deeplinks and app links (`wordpress.com`, `public-api.wordpress.com` tracking links, `wordpress://...` app links) 
- Check the `deep_linked` event in Logcat
- Notice it has the following properties:
  - intent_host - it was there before and is the host of the original link
  - source
  - url (which now contains the stripped URL)
  - source_info (only there for the `Create a site` links if `login_reason` is present in the original tracking URL)

## Regression Notes
1. Potential unintended areas of impact
- All the deeplinks handling

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- I've tested all of the deeplinks I could think of and covered everything with unit tests

3. What automated tests I added (or what prevented me from doing so)
- I've added a ton of unit tests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/deeplinks/EditorLinkHandler.kt [hunk 1]
```diff
@@ -50,11 +54,44 @@ class EditorLinkHandler
...
-    fun isEditorUrl(uri: UriWrapper): Boolean {
+    override fun shouldHandleUrl(uri: UriWrapper): Boolean {
         return (uri.host == DeepLinkingIntentReceiverViewModel.HOST_WORDPRESS_COM &&
                 uri.pathSegments.firstOrNull() == POST_PATH) || uri.host == POST_PATH
     }
 
+    override fun stripUrl(uri: UriWrapper): String {
+        return if (uri.host == POST_PATH) {
+            // Transforms wordpress://post?blogId=798&postId=1231 to wordpress://post?blogId=blogId&postId=postId
+            buildString {
```
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/deeplinks/EditorLinkHandler.kt:67]: TIL about `buildString`!

File: WordPress/src/main/java/org/wordpress/android/ui/deeplinks/NotificationsLinkHandler.kt [hunk 2]
```diff
@@ -6,24 +6,28 @@ import org.wordpress.android.util.UriWrapper
...
      * The handled links are `https://wordpress.com/notifications`
      */
-    fun isNotificationsUrl(uri: UriWrapper): Boolean {
+    override fun shouldHandleUrl(uri: UriWrapper): Boolean {
         return (uri.host == DeepLinkingIntentReceiverViewModel.HOST_WORDPRESS_COM &&
                 uri.pathSegments.firstOrNull() == NOTIFICATIONS_PATH) || uri.host == NOTIFICATIONS_PATH
     }
 
+    override fun stripUrl(uri: UriWrapper): String {
+        return uri.host + "/" + uri.pathSegments.firstOrNull()
```
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/deeplinks/NotificationsLinkHandler.kt]: This is missing the handling of the `wordpress://` scheme. Also, when the path segment is null, like on `wordpress://notifications`, we get an actual `null` appended to the end of the URL. Maybe we should handle this like we do in `StatsLinkHandler`?

**Review by renanferrari — COMMENTED** [review]
Thanks a lot for the changes, the code has definitely improved!

I've tested with a bunch of different links and everything seems to be working as expected, except for one case, which I've detailed below.

Other than that, I thought maybe we should move all link handlers and its related classes (`DeepLinkHandler` and `DeepLinkHandlers`) to their own package inside `deeplinks`, what do you think?

File: WordPress/src/main/java/org/wordpress/android/ui/deeplinks/NotificationsLinkHandler.kt (hunk 2 above)
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/deeplinks/NotificationsLinkHandler.kt]: ah you're right 👍 . I missed this usecase, it's fixed in 
0fe5156

**Comment by planarvoid** [comment]
thanks for the review @renanferrari ! I've fixed the issue and also moved the handlers into an inner package in dee1f57, I think it makes a lot of sense 👍

**Review by renanferrari — APPROVED** [review]
Thanks for the changes!

LGTM :shipit:

**Pull request merged**